### DISCUSSION TITLE
Supporting glob pattern exclusion in route rules

### DISCUSSION THREAD

**Discussion by trc-mathieu** [nitrojs/nitro#2413] [General]
Hi, I have a use case where an api exists at `/api/auth/` and some other endpoints starting with exactly `/api/...`.

I would like to define an exclusion pattern to be able to proxy requests that are not `/api/auth` to somewhere else. I haven't found a way and maybe it's not even possible. I was wondering if you had any suggestions or thoughts on this.
Here's the part where you deal with routeRules: 
https://github.com/unjs/nitro/blob/aabdc9c5c78c011a212c754058160ed23391361d/src/runtime/route-rules.ts#L46-L62

Without exclusion support, I need to explicitly set all the routes as proxy. 

Ideally, I would like something like this to be supported:

```
routeRules:{
 'api/(!auth)**': {proxy: 'https://some-proxied-url/**}
}
```
Right now, I need to do specify every route, which is really cumbersome
```
routeRules:{
 'api/abc/': {proxy: 'https://some-proxied-url/abc/},
 'api/abc/**': {proxy: 'https://some-proxied-url/abc/**},
 'api/bhi/**': {proxy: 'https://some-proxied-url/bhi/**}
[...]
}
```
Thank you for any help on this!

**Comment by pi0** [comment]
Hi. Instead of single `api/(!auth)**`, you should be able to simply use something like this to exclude `api/auth/**` pattern.

```js
routeRules:{
 'api/auth/**': { proxy: false },
 'api/**': { proxy: 'https://some-proxied-url/**' }
}
```

Supporting regex and exclude patterns, while possible and a nice idea for this usecase, will add to generic compatibility and complexity of route rules (that needs to be integratable with cross deployment compatibility) so it is unlikely we add this syntax support.

**Reply by trc-mathieu** [reply]
It would have been great but it doesn't work. The `api/auth/**` is still proxied to `'https://some-proxied-url/**'`. I also tried with leading / . I'm using Nitro 2.9.6

**Reply by trc-mathieu** [reply]
I'm also using radix3 1.1.2, as I understand the route matcher comes from this library

**Reply by pi0** [reply]
It might be a bug then. Do you mind to open an issue with minimal reproduction? 🙏🏼 (above example)

**Reply by trc-mathieu** [reply]
will do, thank you!

**Reply by guibulator** [reply]
@pi0 Here's a reproduction of the issue
https://github.com/guibulator/nitro-route-rules-repro

**Reply by TheAlexLichter** [reply]
@guibulator best to open an issue with that repro + a good description :)

**Reply by guibulator** [reply]
Hum, so it seems to be working with nitro only as you described. Do you know if Nuxt is doing anything special with the routeRules before passing them over to Nitro?
So in this attempt to reproduce the issue with the nitro code sandbox, it's working
https://codesandbox.io/p/github/unjs/nitro/csb-s5h2y3/draft/trusting-blackwell?file=%2Fnitro.config.ts%3A9%2C4

Navigating to https://s5h2y3-3000.csb.app/api/objects does proxy the request and navigating to https://s5h2y3-3000.csb.app/api/auth/info/ doesn't.

**Comment by JavascriptMick** [comment]
@guibulator @TheAlexLichter @trc-mathieu  +1 for this.  I am also struggling with how to produce a v2 server api which 'falls back' to an original default implementation for unchanged endpoints.  It's the sort of thing that should be easy but just does not seem to be possible in an elegant way.

to fill out my e.g, say I have the following handler files...
```
myapp/server/api/stable.ts
myapp/server/api/unstable.ts
myapp/server/api/v2/unstable.ts
```

the behaviour should be pretty obvious

navigating to ...
http://mysite/api/stable - should use the original handler for stable
http://mysite/api/v2/stable - should 'fall back' to the original handler for stable because there is no v2 handler
http://mysite/api/unstable - should use the original handler for unstable
http://mysite/api/v2/unstable - should use the new v2 handler for unstable

so I can *nearly* achieve this with nitro route handler rules...

```
  nitro: {
    routeRules: {
      '/api/v2/**': {
        proxy: '/api/**'
      }
    }
  }
```
OR Nuxt route handlers
```
  routeRules: {
    '/api/v2/**': { proxy: '/api/**' }
  }
```
But it doesn't work properly for http://mysite/api/v2/unstable which proxies back to the old version of the unstable handler..
Effectively, I need to use a glob pattern for the proxy but exclude handlers which have changed...

this doesn't work...
```
  nitro: {
    routeRules: {
      ['/api/v2/(!unstable)**']: {
        proxy: '/api/**'
      }
    }
  }
```

and this also doesn't work

```
  nitro: {
    routeRules: {
      '/api/v2/unstable': {},
      ['/api/v2/**']: {
        proxy: '/api/**'
      }
    }
  }
```
in short, when specifying proxy with globs, there is no way to exclude specific paths from the proxy.
I have also tried a few programattic approaches e.g using a [...].ts and a middleware handler but I couldn't figure it out

This works 
```
  routeRules: {
    '/api/v2/stable': { proxy: '/api/stable' },
    '/api/v2/alsostable': { proxy: '/api/alsostable' }
  },
```
but it's *super* ugly because I need to repeat the proxy for every handler that is not changed in V2.

Is there any better way around this?
